src/random_ray/flat_source_domain.cpp

```diff
@@ -247,12 +252,39 @@ double FlatSourceDomain::compute_k_eff(double k_eff_old) const
       sr_fission_source_new += nu_sigma_f * scalar_flux_new_[idx];
     }
 
+    // Temporary FSR fission rate for performance.
+    double temp_sr_fission_rate = sr_fission_source_new * volume;
+
+    // Calculating old and new fission rates.
     fission_rate_old += sr_fission_source_old * volume;
-    fission_rate_new += sr_fission_source_new * volume;
+    fission_rate_new += temp_sr_fission_rate;
+
+    // Assigning the fission source to the vector for entropy calculation.
+    p[sr] = temp_sr_fission_rate;
   }
 
   double k_eff_new = k_eff_old * (fission_rate_new / fission_rate_old);
 
+  double H = 0.0;
+  // defining an inverse sum for better performance
+  double inverse_sum = 1 / fission_rate_new;
+  // defining inverse log(2) for better performance
+  const double inv_log2 = 1.0 / std::log(2.0);
+
+#pragma omp parallel for reduction(+ : H)
+  for (int i = 0; i < n_source_regions_; i++) {
+    // Only if FSR has fission source
+    if (p[i] != 0.0f) {
```

Comment:
Slightly negative fluxes can happen occasionally (due to the simulation averaged volume estimator), which might cause H to go `NaN`. Might be better to check for greater than zero?
```suggestion
    if (p[i] > 0.0f) {
```

Reply:
Great to know, I've gone ahead and changed this.